app/src/main/java/com/shakenbeer/nutrition/ui/DayListFragment.java

```diff
@@ -187,21 +193,39 @@ protected void onPreExecute() {
             dialog.show();
         }
 
+        private boolean externalStorageAvailable() {
+            return
+                    Environment.MEDIA_MOUNTED
+                            .equals(Environment.getExternalStorageState());
+        }
+
         @Override
         protected Boolean doInBackground(Void... params) {
+
+            File path = null;
+
             SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm");
 
             dir = Environment.DIRECTORY_DOWNLOADS;
             if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
                 dir = Environment.DIRECTORY_DOCUMENTS;
             }
-            File directory = Environment.getExternalStoragePublicDirectory(dir);
-            if (!directory.exists())
-                directory.mkdirs();
-            File file = new File(directory, CARBCULATOR_REPORT_CSV);
 
+            if (externalStorageAvailable()) {
+                path = Environment.getExternalStoragePublicDirectory(dir);
+            } else {
+                path = Environment.getDataDirectory();
```

Comment:
getDataDirectory may not have permissions to read/write and generally work with this directory could cause problems (sometimes you can't see it on computer when connect). Another option could be to show error, if ExternalStorage unavailable. I'm not sure what approach is better, These just thoughts, let's stick with your approach.